File: app/src/main/java/uz/mayasoft/marta/wallet/pin2/security/PFFingerprintPinCodeHelper.java
package uz.mayasoft.marta.wallet.pin2.security;

import android.content.Context;

import androidx.core.hardware.fingerprint.FingerprintManagerCompat;

import uz.mayasoft.marta.wallet.pin2.security.callbacks.PFPinCodeHelperCallback;

public class PFFingerprintPinCodeHelper implements IPFPinCodeHelper {


    private static final String FINGERPRINT_ALIAS = "fp_fingerprint_lock_screen_key_store";
    private static final String PIN_ALIAS = "fp_pin_lock_screen_key_store";

    private static final PFFingerprintPinCodeHelper ourInstance = new PFFingerprintPinCodeHelper();

    public static PFFingerprintPinCodeHelper getInstance() {
        return ourInstance;
    }

    private final IPFSecurityUtils pfSecurityUtils
            = PFSecurityUtilsFactory.getPFSecurityUtilsInstance();

    private PFFingerprintPinCodeHelper() {

    }

    @Override
    public void encodePin(Context context, String pin, PFPinCodeHelperCallback<String> callback) {
        try {
            final String encoded = pfSecurityUtils.encode(context, PIN_ALIAS, pin, false);
            if (callback != null) {
                callback.onResult(new PFResult(encoded));
            }
        } catch (PFSecurityException e) {
            if (callback != null) {
                callback.onResult(new PFResult(e.getError()));
            }
        }
    }

    @Override
    public void checkPin(Context context, String encodedPin, String pin, PFPinCodeHelperCallback<Boolean> callback) {
        try {
            final String pinCode = pfSecurityUtils.decode(PIN_ALIAS, encodedPin);
            if (callback != null) {
                callback.onResult(new PFResult(pinCode.equals(pin)));
            }
        } catch (PFSecurityException e) {
            if (callback != null) {
                callback.onResult(new PFResult(e.getError()));
            }
        }
    }


    private boolean isFingerPrintAvailable(Context context) {
        return FingerprintManagerCompat.from(context).isHardwareDetected();
    }

    private boolean isFingerPrintReady(Context context) {
        return FingerprintManagerCompat.from(context).hasEnrolledFingerprints();
    }

    @Override
    public void delete(PFPinCodeHelperCallback<Boolean> callback) {
        try {
            pfSecurityUtils.deleteKey(PIN_ALIAS);
            if (callback != null) {
                callback.onResult(new PFResult(true));
            }
        } catch (PFSecurityException e) {
            if (callback != null) {
                callback.onResult(new PFResult(e.getError()));
            }
        }
    }

    @Override
    public void isPinCodeEncryptionKeyExist(PFPinCodeHelperCallback<Boolean> callback) {
        try {
            final boolean isExist = pfSecurityUtils.isKeystoreContainAlias(PIN_ALIAS);
            if (callback != null) {
                callback.onResult(new PFResult(isExist));
            }
        } catch (PFSecurityException e) {
            if (callback != null) {
                callback.onResult(new PFResult(e.getError()));
            }
        }
    }

}

File: app/src/main/java/com/example/marta/ui/pin/security/PFFingerprintPinCodeHelper.kt
//package com.example.marta.ui.pin.security
//
//import android.content.Context
//import androidx.core.hardware.fingerprint.FingerprintManagerCompat
//
//import com.example.marta.ui.pin.security.calbacs.PFPinCodeHelperCallback
//
//class PFFingerprintPinCodeHelper private constructor() : IPFPinCodeHelper {
//    private val pfSecurityUtils: IPFSecurityUtils =
//        PFSecurityUtilsFactory.pFSecurityUtilsInstance
//
//    private fun isFingerPrintAvailable(context: Context): Boolean {
//        return FingerprintManagerCompat.from(context)
//            .isHardwareDetected
//    }
//
//    private fun isFingerPrintReady(context: Context): Boolean {
//        return FingerprintManagerCompat.from(context)
//            .hasEnrolledFingerprints()
//    }
//
//
//
//    companion object {
//        private const val FINGERPRINT_ALIAS = "fp_fingerprint_lock_screen_key_store"
//        private const val PIN_ALIAS = "fp_pin_lock_screen_key_store"
//        val instance = PFFingerprintPinCodeHelper()
//    }
//
//    override fun encodePin(
//        context: Context?,
//        pin: String?,
//        callBack: PFPinCodeHelperCallback<String>
//    ) {
//        try {
//            val encoded = pfSecurityUtils.encode(
//                context!!,
//                PIN_ALIAS,
//                pin,
//                false
//            )
//            callBack.onResult(PFResult(encoded))
//        } catch (e: PFSecurityException) {
//            callBack.onResult(PFResult(e.error))
//        }
//    }
//
//    override fun checkPin(
//        context: Context?,
//        encodedPin: String?,
//        pin: String?,
//        callback: PFPinCodeHelperCallback<PFResult<Boolean>>
//    ) {
//        try {
//            val pinCode =
//                pfSecurityUtils.decode(PIN_ALIAS, encodedPin)
//            callback.onResult(PFResult(pinCode.equals(pin)))
//        } catch (e: PFSecurityException) {
//            callback?.onResult(PFResult(e.error))
//        }
//    }
//
//    override fun delete(callback: PFPinCodeHelperCallback<PFResult<Boolean>>) {
//        try {
//            pfSecurityUtils.deleteKey(PIN_ALIAS)
//            callback?.onResult(PFResult(true))
//        } catch (e: PFSecurityException) {
//            callback?.onResult(PFResult(e.error))
//        }
//    }
//
//    override fun isPinCodeEncryptionKeyExist(callback: PFPinCodeHelperCallback<PFResult<Boolean>>) {
//        try {
//            val isExist =
//                pfSecurityUtils.isKeystoreContainAlias(PIN_ALIAS)
//            callback?.onResult(PFResult(isExist))
//        } catch (e: PFSecurityException) {
//            callback?.onResult(PFResult(e.error))
//        }
//    }
//}
//
